File: src/main/java/com/example/chatService/controller/RedisController.java
package com.example.chatService.controller;

import com.example.chatService.dto.*;
import com.example.chatService.redis.OnlineStatusService;
import lombok.RequiredArgsConstructor;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.simp.SimpMessageHeaderAccessor;
import org.springframework.web.bind.annotation.*;

@RestController
@RequiredArgsConstructor
public class RedisController {

    private final OnlineStatusService onlineStatusService;

    @MessageMapping("/user.enter")
    public void userEnter(UserEnterDto dto) {
        if (dto.getUserId() == null || dto.getUsername() == null) return;
        onlineStatusService.markOnline(dto);
    }
    @MessageMapping("/user.heartbeat")
    public void heartbeat(SimpMessageHeaderAccessor accessor) {
        String userId = (String) accessor.getSessionAttributes().get("userId");
        if ( userId != null) {
            onlineStatusService.refreshTTL(userId);
        }
    }

    @MessageMapping("/user.leave")
    public void leave(SimpMessageHeaderAccessor accessor) {
        String userId = (String) accessor.getSessionAttributes().get("userId");
        if ( userId != null) {
            onlineStatusService.markOffline(userId);
        }
    }
}

File: src/main/java/com/example/chatService/controller/RedisController.kt
package com.example.chatService.controller

import com.example.chatService.dto.UserEnterDto
import com.example.chatService.redis.OnlineStatusService
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.messaging.simp.SimpMessageHeaderAccessor
import org.springframework.web.bind.annotation.RestController

@RestController
class RedisController(
    private val onlineStatusService: OnlineStatusService,
) {
    @MessageMapping("/user.enter")
    fun userEnter(dto: UserEnterDto) {
        if (dto.userId == null || dto.username == null) {
            return
        }
        onlineStatusService.markOnline(dto)
    }

    @MessageMapping("/user.heartbeat")
    fun heartbeat(accessor: SimpMessageHeaderAccessor) {
        val userId = accessor.sessionAttributes?.get("userId") as? String ?: return
        onlineStatusService.refreshTTL(userId)
    }

    @MessageMapping("/user.leave")
    fun leave(accessor: SimpMessageHeaderAccessor) {
        val userId = accessor.sessionAttributes?.get("userId") as? String ?: return
        onlineStatusService.markOffline(userId)
    }
}
